BUGFIX: Fix hidden state evaluation

mhsdesign opened the pull request:
The Neos 8.3 LinkingService contained this logic
```
$action = $node->getContext()->getWorkspace()->isPublicWorkspace() && !$node->isHidden() ? 'show' : 'preview';
```
ensuring that disabled nodes can still be previewed. In Neos 9 a no route matched error will be thrown.

To restore the old behaviour we evaluate the hidden state

<!--
Thanks for your contribution, we appreciate it!

ATTENTION: ALL NEW FEATURE PRs SHOULD TARGET THE MASTER BRANCH (bugfixes go to the least maintained branch)
-->


**What I did**

**How I did it**

**How to verify it**

<!--
If possible, a screenshot or a gif comparing the new and old behavior would be great.
-->

mhsdesign committed:
TASK: Fix hidden state evaluation

The Neos 8.3 LinkingService contained this logic
```
$action = $node->getContext()->getWorkspace()->isPublicWorkspace() && !$node->isHidden() ? 'show' : 'preview';
```
ensuring that disabled nodes can still be previewed.
In Neos 9 a no route matched error will be thrown.

To restore the old behaviour we evaluate the hidden state

bwaidelich commented on Classes/Controller/BackendController.php:
```diff
@@ -226,9 +228,29 @@ public function redirectToAction(string $node): void
+
+        $nodeAddressInBaseWorkspace = NodeAddress::create(
+            $nodeAddress->contentRepositoryId,
+            $workspace->baseWorkspaceName ?? WorkspaceName::forLive(),
```
If the selected workspace has no base workspace – shouldn't that be an error?

mhsdesign replied:
i this is like one of these `if ($siteNode) {}` calls that are just there to prevent an error than to run into a fatal error.
If some logic is changed that we for example work on the live workspace directly we should also still be able to preview it ... but i think youre right ... for that case $workspace->baseWorkspaceName ?? $workspace->workspaceName should be enough;)

mhsdesign replied:
Adjusted:)

bwaidelich reviewed (commented):
Thanks for fixing!
The amount of domain logic in the controller makes me a little unsure, but in general it makes sense to me..
Will we be able to cover this with a test to prevent this from breaking again in the future?

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into bugfix/fix-redirectToAction-for-disabled-nodes-in-base-live-workspace

mhsdesign committed:
TASK: Harden `redirectToAction`

remove logic of falling back to live workspace, but use the original workspace then instead.

bwaidelich commented on Classes/Controller/BackendController.php:
```diff
@@ -225,9 +226,29 @@ public function redirectToAction(string $node): void
+
+        $workspace = $contentRepository->findWorkspaceByName($nodeAddress->workspaceName);
+
+        // we always want to redirect to the node in the base workspace.
```
```suggestion
        // we always want to redirect to the node in the base workspace unless we are on a root workspace in which case we stay on that (currently that will not happen)
```

mhsdesign committed:
update comment

Co-authored-by: Bastian Waidelich <b.waidelich@wwwision.de>
